bazelbuild/rules_kotlin
# Implement Jdeps using K2 APIs

This PR moves the kotlin_rules to Kotlin 2.0.0 & adds support for running Jdeps when compiling with K2 (#843).

High level list of changes:
- Enable K2 tests in paramaterized test class
- Compile rules_kotlin with Kotlin 2.0 since we require fixes that were made that aren't in K2 preview of v1.9.23
- Add `JdepsGenExtension2` which leverages `FirAdditionalCheckers` plugin extension to collect info from the FIR tree
- We use the `ClassUsageRecorder` to manage collection of AST info
- The deps info in `ClassUsageRecorder` is init in the registrar & dumped to a proto on the `ClassFileFactoryFinalizerExtension` callback since there is no longer a `onAnalysisCompleted` callback in K2
- Deps info is collected through various declaration & expression checkers using 3 main types `FirClassSymbol`, `FirTypeRef`, & `ConeKotlinType` which is then pushed into the `ClassUsageRecorder` so that supertypes & type args can also be collected, if necessary
- Update the `JdepsGenComponentRegistrar` to invoke K1 or K2 impl based on language version set in the kotlinc configuration
- Override the Kotlin language version to 1.9 for KSP task since KSP does not support 2.0
- Update example apps to work with K2
